File: shared/ios/ReleafCore/Sources/ReleafCoreScan/BusinessCard/Classifier/UrlClassifier.swift
/*
 * UrlClassifier.swift  — mirror of UrlClassifier.kt.
 */

import Foundation

public struct UrlClassifier: Classifier {

    public init() {}

    public func classify(layout: LayoutContext, weights: ScoringWeights) -> [FieldCandidate] {
        var out: [FieldCandidate] = []
        for (idx, block) in layout.blocks.enumerated() {
            let text = block.text
            let nsText = text as NSString
            let fullRange = NSRange(location: 0, length: nsText.length)

            // Pass 1: scheme.
            for m in Self.schemeRegex.matches(in: text, range: fullRange) {
                let raw = nsText.substring(with: m.range)
                out.append(FieldCandidate(
                    sourceBlockIndex: idx,
                    text:             cleanTrailing(raw),
                    kind:             .website,
                    score:            weights.websiteBase + engineConfidenceBonus(block, weights)
                ))
            }

            // Pass 2: www.
            for m in Self.wwwRegex.matches(in: text, range: fullRange) {
                let raw = nsText.substring(with: m.range)
                let withoutWww = String(raw.dropFirst(4))
                if text.contains("://" + withoutWww) { continue }
                out.append(FieldCandidate(
                    sourceBlockIndex: idx,
                    text:             cleanTrailing(raw),
                    kind:             .website,
                    score:            weights.websiteBase + engineConfidenceBonus(block, weights)
                ))
            }

            // Pass 3: bare domain with known TLD.
            for m in Self.bareDomainRegex.matches(in: text, range: fullRange) {
                guard m.numberOfRanges >= 3 else { continue }
                let tldRange = m.range(at: 2)
                let tld = nsText.substring(with: tldRange).lowercased()
                if !Self.commonTlds.contains(tld) { continue }
                // Reject if preceded by '@' — that's an email host.
                let start = m.range.location
                if start > 0 {
                    let prev = nsText.substring(with: NSRange(location: start - 1, length: 1))
                    if prev == "@" { continue }
                }
                let raw = nsText.substring(with: m.range)
                out.append(FieldCandidate(
                    sourceBlockIndex: idx,
                    text:             cleanTrailing(raw).lowercased(),
                    kind:             .website,
                    score:            (weights.websiteBase - 1.0) + engineConfidenceBonus(block, weights)
                ))
            }
        }
        return out
    }

    private static let schemeRegex: NSRegularExpression = {
        // swiftlint:disable:next force_try
        try! NSRegularExpression(pattern: #"https?://\S+"#, options: [.caseInsensitive])
    }()

    private static let wwwRegex: NSRegularExpression = {
        // swiftlint:disable:next force_try
        try! NSRegularExpression(
            pattern: #"www\.[A-Za-z0-9\-._~]+\.[A-Za-z]{2,24}(?:/\S*)?"#,
            options: [.caseInsensitive]
        )
    }()

    private static let bareDomainRegex: NSRegularExpression = {
        // swiftlint:disable:next force_try
        try! NSRegularExpression(
            pattern: #"\b([A-Za-z0-9](?:[A-Za-z0-9\-]*[A-Za-z0-9])?(?:\.[A-Za-z0-9\-]+)*)\.([A-Za-z]{2,24})\b"#
        )
    }()

    private static let commonTlds: Set<String> = [
        "com", "co", "in", "org", "net", "io", "app", "ai", "dev",
        "gov", "edu", "biz", "info", "us", "uk", "de", "fr", "jp",
        "ca", "au", "br", "ru", "cn", "tech", "design", "studio",
        "store", "online", "site", "xyz", "me",
    ]

    private func cleanTrailing(_ s: String) -> String {
        var out = s
        while let last = out.last, ".,;:)]".contains(last) { out.removeLast() }
        return out
    }
}

File: shared/android/shared/scan/src/main/kotlin/app/releaf/shared/scan/businesscard/classifier/UrlClassifier.kt
/*
 * UrlClassifier.kt
 *
 * Pulls website URLs out of business-card OCR text. Three accept
 * paths, in priority order:
 *   1. Explicit `http://` / `https://` schemes — high confidence.
 *   2. `www.` prefix — also high confidence.
 *   3. Bare-domain heuristic: `<word>.<word>` where the trailing
 *      token is a known TLD (com / in / co / org / net / io / app
 *      / gov / edu / etc.). Lower confidence; emitted with a small
 *      penalty so a stray "abc.xyz" string isn't mistaken for a URL.
 *
 * Output is the URL trimmed of trailing punctuation, lower-cased
 * for the host, scheme preserved when present.
 */

package app.releaf.shared.scan.businesscard.classifier

import app.releaf.shared.scan.businesscard.FieldCandidate
import app.releaf.shared.scan.businesscard.FieldKind
import app.releaf.shared.scan.businesscard.ScoringWeights
import app.releaf.shared.scan.businesscard.pipeline.LayoutContext
import app.releaf.shared.scan.businesscard.pipeline.engineConfidenceBonus

class UrlClassifier : Classifier {

    override fun classify(layout: LayoutContext, weights: ScoringWeights): List<FieldCandidate> {
        val out = mutableListOf<FieldCandidate>()
        for ((idx, block) in layout.blocks.withIndex()) {
            // Pass 1: explicit scheme.
            for (match in SCHEME_URL_REGEX.findAll(block.text)) {
                out += FieldCandidate(
                    sourceBlockIndex = idx,
                    text             = cleanTrailing(match.value),
                    kind             = FieldKind.WEBSITE,
                    score            = weights.websiteBase + engineConfidenceBonus(block, weights),
                )
            }
            // Pass 2: www-prefixed.
            for (match in WWW_URL_REGEX.findAll(block.text)) {
                // Skip if already covered by the scheme pass.
                if (block.text.indexOf("://${match.value.removePrefix("www.")}") >= 0) continue
                out += FieldCandidate(
                    sourceBlockIndex = idx,
                    text             = cleanTrailing(match.value),
                    kind             = FieldKind.WEBSITE,
                    score            = weights.websiteBase + engineConfidenceBonus(block, weights),
                )
            }
            // Pass 3: bare domain with a known TLD.
            for (match in BARE_DOMAIN_REGEX.findAll(block.text)) {
                val tld = match.groupValues[2].lowercase()
                if (tld !in COMMON_TLDS) continue
                // Skip emails — the local part of an email contains
                // an @ which the bare-domain regex doesn't match,
                // but the host part (after @) does. Reject when the
                // match is preceded by "@".
                val start = match.range.first
                if (start > 0 && block.text[start - 1] == '@') continue
                out += FieldCandidate(
                    sourceBlockIndex = idx,
                    text             = cleanTrailing(match.value).lowercase(),
                    kind             = FieldKind.WEBSITE,
                    // Slight discount vs. scheme/www since the signal is weaker.
                    score            = (weights.websiteBase - 1.0) + engineConfidenceBonus(block, weights),
                )
            }
        }
        return out
    }

    companion object {
        private val SCHEME_URL_REGEX = Regex("""https?://\S+""", RegexOption.IGNORE_CASE)
        private val WWW_URL_REGEX    = Regex("""www\.[A-Za-z0-9\-._~]+\.[A-Za-z]{2,24}(?:/\S*)?""", RegexOption.IGNORE_CASE)
        private val BARE_DOMAIN_REGEX = Regex("""\b([A-Za-z0-9](?:[A-Za-z0-9\-]*[A-Za-z0-9])?(?:\.[A-Za-z0-9\-]+)*)\.([A-Za-z]{2,24})\b""")

        private val COMMON_TLDS = setOf(
            "com", "co", "in", "org", "net", "io", "app", "ai", "dev",
            "gov", "edu", "biz", "info", "us", "uk", "de", "fr", "jp",
            "ca", "au", "br", "ru", "cn", "tech", "design", "studio",
            "store", "online", "site", "xyz", "me",
        )

        private fun cleanTrailing(s: String): String =
            s.trimEnd { it == '.' || it == ',' || it == ';' || it == ':' || it == ')' || it == ']' }
    }
}
